server/src/main/java/org/elasticsearch/node/Node.java

```diff
@@ -580,20 +597,93 @@ public synchronized void close() throws IOException {
+
+        @SuppressWarnings(value = "rawtypes") // Can't make an array of parameterized types, but it complains if you leave the type out
+        CompletableFuture<Void> allStoppers = CompletableFuture.allOf(futures.values().toArray(new CompletableFuture[stoppers.size()]));
+
+        TimeValue maxTimeout = MAXIMUM_SHUTDOWN_TIMEOUT_SETTING.get(this.settings());
         try {
-            stopper.get();
-        } catch (Exception e) {
-            logger.warn("unexpected exception while waiting for http server to close", e);
+            if (maxTimeout.millis() == 0) {
+                FutureUtils.get(allStoppers);
+            } else {
+                FutureUtils.get(allStoppers, maxTimeout.millis(), TimeUnit.MILLISECONDS);
+            }
+
+        } catch (ElasticsearchTimeoutException t) {
+            var unfinishedTasks = futures.entrySet()
+                .stream()
+                .filter(entry -> entry.getValue().isDone() == false)
+                .map(Map.Entry::getKey)
+                .toList();
+            logger.info("timed out while waiting for graceful shutdown tasks: " + unfinishedTasks);
+        }
+    }
+
+    private void awaitSearchTasksComplete() {
+        TaskManager taskManager = injector.getInstance(TransportService.class).getTaskManager();
+        TimeValue asyncSearchTimeout = SHUTDOWN_SEARCH_TIMEOUT_SETTING.get(this.settings());
+        long millisWaited = 0;
+        while (true) {
+            long searchTasksRemaining = taskManager.getTasks()
+                .values()
+                .stream()
+                .filter(task -> TransportSearchAction.TYPE.name().equals(task.getAction()))
+                .count();
+            if (searchTasksRemaining == 0) {
+                logger.debug("all search tasks complete");
+                return;
+            } else {
+                // Let the system work on those searches for a while. We're on a dedicated thread to manage app shutdown, so we
+                // literally just want to wait and not take up resources on this thread for now. Poll period chosen to allow short
+                // response times, but checking the tasks list is relatively expensive, and we don't want to waste CPU time we could
+                // be spending on finishing those searches.
+                final TimeValue pollPeriod = TimeValue.timeValueMillis(500);
+                millisWaited += pollPeriod.millis();
+                if (millisWaited >= asyncSearchTimeout.millis()) {
+                    logger.warn(
+                        format(
+                            "timed out after waiting [%s] for [%d] search tasks to finish",
+                            asyncSearchTimeout.toString(),
+                            searchTasksRemaining
+                        )
+                    );
+                    return;
+                }
+                logger.debug(format("waiting for [%s] search tasks to finish, next poll in [%s]", searchTasksRemaining, pollPeriod));
```

Comment:
I wonder how we might make the remaining tasks visible without being too noisy. If we hit the global timeout, we should be sure to log the number of remaining tasks. Perhaps each task in prepareForClose should have a "get status" or "log status" or something like that that can be called if the task is not complete?

Reply:
I don't disagree but I also want to push back on scope creep in this PR. This is a stopgap measure until we can properly build a framework for handling this kind of deep shutdown handling that we definitely need; in order to implement this kind of status handler I'd have to break it out into its own class and that triggered my "this is getting to big" spidey sense.

That said, I absolutely agree we should handle the timeout centrally as much as possible and just tell the handlers when they've timed out from prepareForClose, as well as ideally having a way to enable periodically logging or querying the status of each handler, when we do build that proper framework.

Reply:
Fair enough, we can followup separately to improve logging more.